File: 250902/기울어진 직사각형/slanted-rectangle.java
import java.util.*;

public class Main {

    static int n;
    static int[][] grid;

    static int getSum(int i, int j, int k, int l) {

        int sum = 0;
        int[] dx = {1, 1, -1, -1};
        int[] dy = {-1, 1, 1, -1};
        int[] moveStep = {k, l, k, l};

        int currX = i;
        int currY = j;

        for (int dir = 0; dir < 4; dir++) {
            for (int step = 0; step < moveStep[dir]; step++) {
                if (
                    0 <= currX && currX < n
                    && 0 <= currY && currY < n
                ) {
                    sum += grid[currX][currY];
                    currX += dx[dir];
                    currY += dy[dir];
                } else {
                    return 0;
                }
            }
        }

        return sum;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        grid = new int[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                grid[i][j] = sc.nextInt();
        // Please write your code here.

        int max = 0;        

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 1; k < n; k++) {
                    for (int l = 1; l < n; l++) {
                        int sum = getSum(i,j,k,l);
                        // if (sum > max) {
                        //     System.out.println(
                        //         i + " " + j + " " + k + " " + l
                        //     );
                        // }
                        max = Math.max(max, sum);             
                    }
                }                       
            }
        }

        System.out.println(max);        
    }
}

File: 260317/기울어진 직사각형/slanted-rectangle.py
n = int(input())
grid = [list(map(int, input().split())) for _ in range(n)]

# Please write your code here.


max_sum = 0
dr = [-1, -1, 1, 1]
dc = [1, -1, -1, 1]

for r in range(n):
    for c in range(n):        
        for w in range(1, n):
            for h in range(1, n):
                sum = 0
                rr = r
                cc = c
                is_valid = True
                for ww in range(w):
                    rr += dr[0]
                    cc += dc[0]
                    if cc < n and rr < n and 0<= cc and 0<= rr:
                        sum += grid[rr][cc]
                    else:
                        is_valid = False
                        break
                for hh in range(h):
                    rr += dr[1]
                    cc += dc[1]
                    if cc < n and rr < n and 0<= cc and 0 <= rr:
                        sum += grid[rr][cc]
                    else:
                        is_valid = False
                        break
                for ww in range(w):
                    rr += dr[2]
                    cc += dc[2]
                    if cc < n and rr < n and 0<= cc and 0<= rr:
                        sum += grid[rr][cc]
                    else:
                        is_valid = False
                        break
                for hh in range(h):
                    rr += dr[3]
                    cc += dc[3]
                    if cc < n and rr < n and 0<= cc and 0 <= rr:
                        sum += grid[rr][cc]
                    else:
                        is_valid = False
                        break

                if is_valid and max_sum < sum:
                    # print(f'{r} {c} {w} {h}')
                    max_sum = max(max_sum, sum)




print(max_sum)

                
         

                
                